Feature/add msaljs

thewahome-zz opened the pull request:
## Overview
Adds Msal as the authentication library and phases out hellojs

### Demo

Optional. Screenshots, `curl` examples, etc.

### Notes

* had to use loginPopup method instead of loginRedirect since the latter does not properly bring the id token back to the callback
* need to pass in the scopes when logging in and getting the access token
* client id temporarily moved to the auth.service.ts file
* storing a status string in local storage and monitoring it to check authentication status

## Testing Instructions

* Enter the appropriate client id in the auth.services.ts file
* send requests as an anonymous user
* log in
* send requests as your logged in self

thewahome committed:
add msal to solution

thewahome committed:
add msal auth service to replace hello js

thewahome committed:
change authentication status after successful login

thewahome committed:
use localstorage to store authentication status

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/signed)](https://cla.opensource.microsoft.com/microsoftgraph/microsoft-graph-explorer?pullRequest=190) <br/>All CLA requirements met.

thewahome committed:
merge in dev branch with tslint changes

thewahome committed:
fix tslint errors

thewahome committed:
follow tslint guides

MIchaelMainer commented on src/app/authentication/auth.service.ts:
```diff
@@ -6,20 +6,20 @@ export class AuthService {
     public login() {
         return this.app.loginPopup(this.applicationConfig.graphScopes)
-            .then(idToken => {
+            .then((idToken) => {
```
Not really part of this PR. We aren't using the idToken. We don't request it.

MIchaelMainer commented on src/app/authentication/authentication.component.ts:
```diff
@@ -1,78 +1,77 @@
           this.createUserProfile();
         } else {
-          console.log('login failed');
+          // Console.log('login failed');
```
Let's delete this and/or provide a proper comment here.

MIchaelMainer commented on src/app/authentication/authentication.component.ts:
```diff
@@ -1,78 +1,77 @@
+      const blob = new Blob([result.arrayBuffer()], { type: 'image/jpeg' });
+      const imageUrl = window.URL.createObjectURL(blob);
       AppComponent.explorerValues.authentication.user.profileImageUrl = imageUrl;
-    }).catch((e) => console.log(e)));
```
General comment: all of the places where we remove the .catch and console.log will be candidates for adding telemetry. We should update backlog item 449602 whenever we have PR where we are removing these .catch then console.log.

MIchaelMainer commented on src/app/authentication/auth.service.ts:
```diff
@@ -0,0 +1,51 @@
+@Injectable()
+export class AuthService {
+    private applicationConfig: any = {
+        clientID: '<Insert clint Id>', //TODO get the client Id from configuration file
```
Address this TODO for getting the clientId.

thewahome-zz replied:
on it...

MIchaelMainer commented on src/app/authentication/auth.service.ts:
```diff
@@ -0,0 +1,51 @@
+    private app: any;
+
+    constructor() {
+        // This.applicationConfig.clientID = AppComponent.Options.ClientId;
```
Remove comment

MIchaelMainer commented on src/app/graph-service.ts:
```diff
@@ -1,71 +1,75 @@
+    constructor(private http: Http, private authService: AuthService) { }
 
-  public performAnonymousQuery(queryType: RequestType, query: string, headers?: Headers): Promise<Response> {
+    performAnonymousQuery(queryType:RequestType, query:string, headers?:Headers):Promise<Response> {
```
General comment. I was under the impression that we were going to explicit access modifiers. Plus, the spacing between parameters has changed. Add access modifiers and make the spacing consistent.

jobala replied:
Yes that is the case.  The rules enforce access modifiers.

MIchaelMainer commented on src/app/graph-service.ts:
```diff
@@ -1,71 +1,75 @@
+            return this.http.get(`https://proxy.apisandbox.msdn.microsoft.com/svc?url=${encodeURIComponent(query)}`, {headers, responseType: ResponseContentType.ArrayBuffer}).toPromise();
         }
     }
+    // let method = isAuthenticated() ? this.GraphService.performQuery : this.GraphService.performAnonymousQuery;
```
remove comment.

MIchaelMainer commented on src/app/graph-service.ts:
```diff
@@ -1,71 +1,75 @@
 
         if (!sentToGraph) {
-            throw new Error('Not sending request to known Graph deployment');
+            throw 'Not sending request to known Graph deployment';
```
We want the message wrapped in an Error object. Especially once we start extending the Error object, it will give more information about the error source for debugging. If we need to provide additional context to a error handler, we should name the error.

MIchaelMainer reviewed (changes requested):
First off, I'm excited to get this running. I'll run this in a bit.

Let's add some indent rules to the .vscode settings since I think we might all be out of sync.

Why do we still have HelloJS code? Does MSAL not cover some scenarios?

MIchaelMainer commented:
We need to get the CI issues resolved as well. We need to run `npm run test` before pushing.

thewahome committed:
set client Id from configuration file

thewahome committed:
fix failing could not find Msal error

MIchaelMainer commented:
On Mac+Chrome, the authentication button was hidden upon load. If I run one of the samples, the authentication button shows. I successfully sign in with my work account.

I can't use the modify permissions UI to add additional scopes. That needs to get hooked up.

darrelmiller commented on src/app/authentication/auth.ts:
```diff
@@ -1,19 +1,20 @@
+export function initAuth(options: IExplorerOptions, apiService: GraphService, changeDetectorRef: ChangeDetectorRef,
+                         authService: AuthService) {
+    setInterval(refreshAccessToken, 1000 * 60 * 10); // Refresh access token every 10 minutes
+    hello.init({
```
Why do we still have references to hello.js here?  I thought this entire auth.ts was being replaced by auth.service,ts.

darrelmiller commented on src/app/authentication/authentication.component.ts:
```diff
@@ -1,58 +1,98 @@
     ScopesDialogComponent.showDialog();
   }
+
+  private createUserProfile() {
```
I would suggest this be called "DisplayUserProfile", "RefreshUserProfile" or "ShowUserProfile".  Something that indicates we are just updating the UI and not actually creating anything.

muthurathinam commented on src/app/authentication/auth.service.ts:
```diff
@@ -0,0 +1,50 @@
+export class AuthService {
+    private applicationConfig: any = {
+        clientID: (window as any).ClientId,
+        graphScopes: ['openid', 'profile', 'User.ReadWrite', 'User.ReadBasic.All', 'Sites.ReadWrite.All',
```
Its better we have some constants file for scopes from which we can use the scopes

muthurathinam commented on package-lock.json:
```diff
@@ -26,8 +26,8 @@
       "dev": true,
       "requires": {
         "@angular/tsc-wrapped": "4.1.3",
-        "minimist": "^1.2.0",
```
Why is this caret symbol removed from versioning ? Yes I understand it is generated file do we have know the reason for this ? This applies to all changes in this file.

thewahome-zz replied:
Not sure.. It was changed automatically on npm install

muthurathinam commented on src/app/authentication/auth.service.ts:
```diff
@@ -0,0 +1,50 @@
+import { Injectable } from '@angular/core';
+declare var Msal: any;
```
Why are we declaring `msal` as `any` here can't we import it ?

thewahome-zz replied:
We are using msal core which does not export modules. Hence the need to declare msal there...

muthurathinam commented on package.json:
```diff
@@ -55,6 +55,7 @@
     "karma-jasmine": "^1.1.0",
     "karma-jasmine-html-reporter": "^0.2.2",
     "moment": "^2.21.0",
+    "msal": "^0.2.3",
```
we still have hellojs dependency ? refer line 51

thewahome-zz replied:
removing hellojs dependency ...

thewahome committed:
merge dev branch changes

thewahome committed:
change log in method to redirect instead of popup

thewahome committed:
change variable names and declarations

muthurathinam commented on src/app/authentication/auth.service.ts:
```diff
@@ -1,5 +1,5 @@
 import { Injectable } from '@angular/core';
-declare var Msal: any;
+declare let Msal: any;
```
Its better we declare it as const.

thewahome committed:
change default scopes format

thewahome committed:
clean up graph service

jobala committed:
Fixes failing tests

jobala committed:
Updates nodejs to version 7 in travis

thewahome committed:
import jwt-decode

thewahome committed:
get and set consented scopes

thewahome committed:
set scopes when logging in

thewahome committed:
prevent logout when fetching user's profile

MIchaelMainer commented on .travis.yml:
```diff
@@ -5,7 +5,7 @@ addons:
   chrome: stable
 node_js:
   - "7"
-  - "node"
```
We might as well remove line 8 instead of running this twice against Node 7. Running against Node latest was useful as it let us know that Node 11 will cause us some problems. But we don't want to block ourselves on that now. So let's remove line 8.

thewahome committed:
change time that set permissions is called

thewahome committed:
Merge branch 'dev' into feature/add-msaljs

thewahome committed:
Merge branch 'feature/add-msaljs' into add-msaljs-scopes

# Conflicts:
#	package-lock.json

thewahome committed:
pass newly selected scopes to getToken method

thewahome committed:
create callback for acquireTokenRedirect

thewahome committed:
use popups instead of redirects

thewahome committed:
remove authentication errors when calling loginPopup

thewahome committed:
removes the display of admin consent alert

thewahome committed:
remove caret from jasmine

thewahome committed:
improves the speed the consent popup appears

thewahome committed:
adds jwt-decode to karma configurations

thewahome committed:
split acquireTokenSilent and acquireTokenPopup

thewahome committed:
checks valid token when app loaded for the first time

thewahome committed:
installs jasmine without caret

thewahome committed:
set jasmine to =2.5.47

thewahome committed:
replace tokenPlease

thewahome committed:
remove hellojs

thewahome committed:
delete history from local storage

thewahome committed:
comment and specify method return types

thewahome committed:
detects changes at different login process

thewahome committed:
changes get token methods declaration

# declaring the parameters explicitly causes the tests to fail

thewahome committed:
merge changes from dev branch

thewahome committed:
clear session storage on logout

thewahome committed:
merge in changes from dev

thewahome committed:
adds missing changes

thewahome committed:
remove hellojs dependency

thewahome committed:
uses async/await

thewahome committed:
use preview msal, get consented scopes

thewahome committed:
fetch updated list of scopes (on reload pending)

thewahome committed:
force page reload on scopes change

thewahome committed:
reference msal in blob storage

thewahome committed:
export authservice functions

jobala committed:
Updates karma dependencies

jobala committed:
This commit is a dirty abomination of a hack.

It included a node_modules folder in src/ since that's where systemjs was loooking for msal during testing.

thewahome committed:
remove valid token check on page load

thewahome committed:
Merge branch 'feature/add-msaljs' of https://github.com/microsoftgraph/microsoft-graph-explorer into feature/add-msaljs

jobala committed:
Adds node_modules

jobala committed:
Linter ignores node_modules in src/

jobala committed:
Adds msal dist folder

thewahome committed:
define graphService in tests

jobala committed:
Merge branch 'feature/add-msaljs' of https://github.com/microsoftgraph/microsoft-graph-explorer into feature/add-msaljs

thewahome committed:
adds msal and systemjs to karma.conf.js

thewahome committed:
change log level

jobala committed:
Merge branch 'feature/add-msaljs' of https://github.com/microsoftgraph/microsoft-graph-explorer into feature/add-msaljs

jobala committed:
Merge branch 'feature/add-msaljs' of https://github.com/microsoftgraph/microsoft-graph-explorer into feature/add-msaljs

jobala committed:
Uses older versions of Karma

thewahome committed:
enables login to canary

darrelmiller reviewed (approved):
I tested locally and canary mode worked for me.

MIchaelMainer reviewed (approved):
Tested locally with node 10.15.3, it works as expected.  I did notice that if I try to select a new permission and then select *Return to the application without granting consent*, it reloads the page. It maybe something beyond our control.

![image](https://user-images.githubusercontent.com/8527305/56774112-0b9ab180-6776-11e9-87b3-cc65ad1f252e.png)
